be/src/formats/orc/orc_chunk_writer.cpp

```diff
@@ -0,0 +1,736 @@
+            OrcTimestampHelper::native_ts_to_orc_ts(values[i], timestamp_orc_column.data[i],
+                                                    timestamp_orc_column.nanoseconds[i]);
+        }
+        orc_column.hasNulls = true;
+    } else {
+        auto* values = ColumnHelper::cast_to_raw<TYPE_DATETIME>(column)->get_data().data();
+        for (size_t i = 0; i < column_size; ++i) {
+            OrcTimestampHelper::native_ts_to_orc_ts(values[i], timestamp_orc_column.data[i],
+                                                    timestamp_orc_column.nanoseconds[i]);
+        }
+        memset(timestamp_orc_column.notNull.data(), 1, column_size * sizeof(char));
+    }
+    timestamp_orc_column.numElements = column_size;
+}
+
+Status OrcChunkWriter::_write_array_column(orc::ColumnVectorBatch& orc_column, ColumnPtr& column,
+                                           const TypeDescriptor& type) {
+    auto& array_orc_column = dynamic_cast<orc::ListVectorBatch&>(orc_column);
+    auto column_size = column->size();
+
+    array_orc_column.resize(column_size);
+    array_orc_column.notNull.resize(column_size);
+    auto& value_orc_column = *array_orc_column.elements;
+
+    if (column->is_nullable()) {
+        auto* col_nullable = down_cast<NullableColumn*>(column.get());
+        auto* array_cols = down_cast<ArrayColumn*>(col_nullable->data_column().get());
+        uint32_t* offsets = array_cols->offsets_column().get()->get_data().data();
+        auto* nulls = col_nullable->null_column()->get_data().data();
+
+        array_orc_column.offsets[0] = offsets[0];
+        for (size_t i = 0; i < column_size; ++i) {
+            array_orc_column.offsets[i + 1] = offsets[i + 1];
+            array_orc_column.notNull[i] = !nulls[i];
+        }
+
+        RETURN_IF_ERROR(_write_column(value_orc_column, array_cols->elements_column(), type.children[0]));
+        array_orc_column.hasNulls = true;
+    } else {
+        auto* array_cols = down_cast<ArrayColumn*>(column.get());
+        uint32_t* offsets = array_cols->offsets_column().get()->get_data().data();
+
+        array_orc_column.offsets[0] = offsets[0];
+        for (size_t i = 0; i < column_size; ++i) {
+            array_orc_column.offsets[i + 1] = offsets[i + 1];
+        }
+        memset(array_orc_column.notNull.data(), 1, column_size * sizeof(char));
+
+        RETURN_IF_ERROR(_write_column(value_orc_column, array_cols->elements_column(), type.children[0]));
+    }
+    array_orc_column.numElements = column_size;
+    return Status::OK();
+}
+
+Status OrcChunkWriter::_write_struct_column(orc::ColumnVectorBatch& orc_column, ColumnPtr& column,
+                                            const TypeDescriptor& type) {
+    auto& struct_orc_column = dynamic_cast<orc::StructVectorBatch&>(orc_column);
+    auto column_size = column->size();
+
+    struct_orc_column.resize(column_size);
+    struct_orc_column.notNull.resize(column_size);
+
+    if (column->is_nullable()) {
```

Comment:
Maybe you need to extract the same logic to make the code more readable in the later PR.

Reply:
Okay